File: Java/src/Primes.java
import java.util.ArrayList;

public class Primes {
    private static boolean isPrime(int n, ArrayList<Integer> primes) {
        double sq = Math.sqrt(n);
        for (int i : primes) {
            if (i > sq) break;
            if (n % i == 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        long start = System.currentTimeMillis();
        ArrayList<Integer> primes = new ArrayList<>(500000);
        for (int i = 2; i < 1000000; i++) {
            if (isPrime(i, primes)) {
                primes.add(i);
            }
        }
        System.out.println(primes.size() + " primes were found");
        System.out.println("Process finished in " + (System.currentTimeMillis() - start) + "ms");
    }
}

File: C/Primes.c
#include <stdio.h>
#include <time.h>
#include <stdbool.h>
#include <math.h>

bool checkPrime(int x, int *primes, int index) {
    double sq = sqrt(x);
    for(int i = 0; i < index && primes[i] <= sq; ++i)  {
        if (x % primes[i] == 0){
            return false;
        }
    }
    return true;
}

int main()
{
    int maxNum = 1000000;
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC_RAW, &start);

    int values[maxNum / 2];
    int index = 0;
    for (int x = 2; x< maxNum; ++x) {
        if (checkPrime(x, values, index)) {
            ++index;
        }
    }
    clock_gettime(CLOCK_MONOTONIC_RAW, &end);
    printf("found %d, \n", index);
    printf("Finished in ms: %ld\n", (end.tv_sec - start.tv_sec) * 1000 + (end.tv_nsec - start.tv_nsec) / 1000000);
}
